cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java

```diff
@@ -0,0 +1,149 @@
+import co.cask.cdap.data2.registry.RuntimeUsageRegistry;
+import co.cask.cdap.explore.guice.ExploreClientModule;
+import co.cask.cdap.internal.app.queue.QueueReaderFactory;
+import co.cask.cdap.internal.app.store.remote.RemoteLineageWriter;
+import co.cask.cdap.internal.app.store.remote.RemoteRuntimeStore;
+import co.cask.cdap.internal.app.store.remote.RemoteRuntimeUsageRegistry;
+import co.cask.cdap.logging.guice.LoggingModules;
+import co.cask.cdap.metrics.guice.MetricsClientRuntimeModule;
+import co.cask.cdap.notifications.feeds.client.NotificationFeedClientModule;
+import com.google.inject.AbstractModule;
+import com.google.inject.Module;
+import com.google.inject.PrivateModule;
+import com.google.inject.Scopes;
+import com.google.inject.assistedinject.FactoryModuleBuilder;
+import com.google.inject.name.Names;
+import com.google.inject.util.Modules;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.api.ServiceAnnouncer;
+import org.apache.twill.api.TwillContext;
+import org.apache.twill.common.Cancellable;
+
+import java.net.InetAddress;
+
+/**
+ * Defines guice modules for distributed program runnables. For instance, AbstractProgramTwillRunnable, as well as
+ * mapreduce tasks / spark executors.
+ */
+public class DistributedProgramRunnableModule {
+
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+
+  public DistributedProgramRunnableModule(CConfiguration cConf, Configuration hConf) {
+    this.cConf = cConf;
+    this.hConf = hConf;
+  }
+
+  // usable from any program runtime, such as mapreduce task, spark task, etc
+  public Module createModule() {
+    Module combined = Modules.combine(
+      new ConfigModule(cConf, hConf),
+      new IOModule(),
+      new ZKClientModule(),
+      new KafkaClientModule(),
+      new MetricsClientRuntimeModule().getDistributedModules(),
+      new LocationRuntimeModule().getDistributedModules(),
+      new LoggingModules().getDistributedModules(),
+      new DiscoveryRuntimeModule().getDistributedModules(),
+      new DataFabricModules().getDistributedModules(),
+      new DataSetsModules().getDistributedModules(),
+      // why do we need explore here?
+      new ExploreClientModule(),
+      new ViewAdminModules().getDistributedModules(),
+      new StreamAdminModules().getDistributedModules(),
+      new NotificationFeedClientModule(),
+      new AuditModule().getDistributedModules(),
+      new AbstractModule() {
+        @Override
+        protected void configure() {
+          // For Binding queue stuff
+          bind(QueueReaderFactory.class).in(Scopes.SINGLETON);
+
+          // For binding DataSet transaction stuff
+          install(new DataFabricFacadeModule());
+
+          bind(RuntimeStore.class).to(RemoteRuntimeStore.class);
+
+          // For binding StreamWriter
+          install(createStreamFactoryModule());
+        }
+      }
+    );
+
+    return Modules.override(combined).with(new AbstractModule() {
```

Comment:
Hmm, I think I bound it in DataSets module, because that's where the bindings are configured for LineageWriter in terms of inmemory, standalone, vs distributed.
This override is a subset of distributed (for program runnables).